File: scala/robot-simulator/src/main/scala/Robot.scala
import Bearing._
import Instruction._
import typeclass._

case class Robot(bearing: Bearing, coordinates: Coordinates) {
  def advance: Robot =
    copy(coordinates = bearing.advance(coordinates))

  // bring it into scope
  import CircularBearing._

  def turnRight(): Robot =
    copy(bearing = bearing.next)

  def turnLeft(): Robot =
    copy(bearing = bearing.previous)

  def simulate(instructions: Instructions): Robot = {
    def execute(robot: Robot, instruction: Instruction): Robot =
      instruction.execute(robot)

    instructions.foldLeft(this)(execute)
  }
}

object Bearing {
  type Coordinates = (Int,Int)

  private type Advance = Coordinates => Coordinates

  sealed trait Bearing {
    def advance: Advance
  }

  case object North extends Bearing {
    override val advance: Advance = { case (x, y) => (x, y + 1) }
  }
  case object South extends Bearing {
    override val advance: Advance = { case (x, y) => (x, y - 1) }
  }
  case object West extends Bearing {
    override val advance: Advance = { case (x, y) => (x - 1, y) }
  }
  case object East extends Bearing {
    override val advance: Advance = { case (x, y) => (x + 1, y) }
  }

  private val values: Seq[Bearing] = Seq(North, East, South, West)

  implicit object BoundedBearing extends Bounded[Bearing] {
    override val minBound = values.head
    override val maxBound = values.last
  }

  implicit object EnumBearing extends Enum[Bearing] {
    override val pred: Bearing => Bearing =
      bearing => values(values.indexOf(bearing) - 1)
    override val succ: Bearing => Bearing =
      bearing => values(values.indexOf(bearing) + 1)
  }

  implicit object CircularBearing extends Circular[Bearing] {
    override val bounded = BoundedBearing
    override val enum = EnumBearing
  }
}

object Instruction {
  type Instruction = Char
  type Instructions = Seq[Instruction]

  implicit class InstructionOps(self: Instruction) {
    def execute(robot: Robot): Robot =
      self match {
        case 'R' => robot.turnRight
        case 'L' => robot.turnLeft
        case 'A' => robot.advance
      }
  }
}

File: haskell/robot-simulator/src/Robot.hs
{-# LANGUAGE DeriveAnyClass  #-}
module Robot (Bearing(..), Robot, mkRobot,
              coordinates, simulate,
              bearing, turnRight, turnLeft) where

type Coordinates = (Integer, Integer)
type Instruction = Char
type Instructions = [Instruction]

class (Enum a, Bounded a, Eq a) => Circular a where
  previous :: a -> a
  previous x
    | x == minBound = maxBound
    | otherwise = pred x

  next :: a -> a
  next x
    | x == maxBound = minBound
    | otherwise = succ x

data Bearing = North | East | South | West
  deriving (Eq, Show, Bounded, Enum, Circular)

data Robot = Robot { bearing :: Bearing, coordinates :: Coordinates }
  deriving (Eq, Show)

simulate :: Robot -> Instructions -> Robot
simulate = foldl execute

execute :: Robot -> Instruction -> Robot
execute (Robot b c@(x, y)) instruction =
  case instruction of
    'L' -> mkRobot (turnLeft b) c
    'R' -> mkRobot (turnRight b) c
    'A' -> mkRobot b advance
    i   -> error $ "invalid instruction: " ++ [i]
  where
    advance :: Coordinates
    advance =
      case b of
        South -> (x, y-1)
        North -> (x, y+1)
        West  -> (x-1, y)
        East  -> (x+1, y)

mkRobot :: Bearing -> Coordinates -> Robot
mkRobot = Robot

turnRight :: Bearing -> Bearing
turnRight = next

turnLeft :: Bearing -> Bearing
turnLeft = previous
